### PR TITLE
IGNITE-13810 append expire policy info into log of started cache

### PR DESCRIPTION
**Pull request by JKonSir**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

**Comment by ivandasch** [comment]
I suppose that constantly running python tests doesn't help.

If you look at logs, you will see NPE 
```
[2021-05-19T11:17:57,828][ERROR][exchange-worker-#118][GridDhtPartitionsExchangeFuture] Failed to initialize cache(s) (will try to rollback) [exchId=GridDhtPartitionExchangeId [topVer=AffinityTopologyVersion [topVer=3, minorTopVer=8], discoEvt=DiscoveryCustomEvent [customMsg=DynamicCacheChangeBatch [id=3f923b38971-7e2b344a-a560-4e74-a44e-8798a26c073a, reqs=ArrayList [DynamicCacheChangeRequest [cacheName=expiry_cache, hasCfg=true, nodeId=b2f50d47-946d-49cd-82f4-a869a3b542bd, clientStartOnly=false, stop=false, destroy=false, disabledAfterStart=false]], exchangeActions=ExchangeActions [startCaches=[expiry_cache], stopCaches=null, startGrps=[expiry_cache], stopGrps=[], resetParts=null, stateChangeRequest=null], startCaches=false], affTopVer=AffinityTopologyVersion [topVer=3, minorTopVer=8], super=DiscoveryEvent [evtNode=TcpDiscoveryNode [id=b2f50d47-946d-49cd-82f4-a869a3b542bd, consistentId=srv_1, addrs=ArrayList [127.0.0.1], sockAddrs=HashSet [/127.0.0.1:48500], discPort=48500, order=1, intOrder=1, lastExchangeTime=1621412227458, loc=true, ver=2.11.0#20210519-sha1:00000000, isClient=false], topVer=3, msgTemplate=null, span=o.a.i.i.processors.tracing.NoopSpan@695cee68, nodeId8=b2f50d47, msg=null, type=DISCOVERY_CUSTOM_EVT, tstamp=1621412277807]], nodeId=b2f50d47, evt=DISCOVERY_CUSTOM_EVT], caches=[o.a.i.i.processors.cache.ExchangeActions$CacheGroupActionData@3b5d0f91]]
java.lang.NullPointerException: null
  at org.apache.ignite.internal.processors.cache.GridCacheProcessor.buildExpirePolicyInfo(GridCacheProcessor.java:2410) ~[classes/:?]
  at org.apache.ignite.internal.processors.cache.GridCacheProcessor.onCacheStarted(GridCacheProcessor.java:2311) ~[classes/:?]
  at org.apache.ignite.internal.processors.cache.GridCacheProcessor.prepareCacheStart(GridCacheProcessor.java:1977) ~[classes/:?]
  at org.apache.ignite.internal.processors.cache.GridCacheProcessor.lambda$prepareStartCaches$55a0e703$1(GridCacheProcessor.java:1841) ~[classes/:?]
  at org.apache.ignite.internal.processors.cache.GridCacheProcessor.lambda$prepareStartCachesIfPossible$14(GridCacheProcessor.java:1811) ~[classes/:?]
  at org.apache.ignite.internal.processors.cache.GridCacheProcessor.prepareStartCaches(GridCacheProcessor.java:1838) ~[classes/:?]
  at org.apache.ignite.internal.processors.cache.GridCacheProcessor.prepareStartCachesIfPossible(GridCacheProcessor.java:1809) ~[classes/:?]
  at org.apache.ignite.internal.processors.cache.CacheAffinitySharedManager.processCacheStartRequests(CacheAffinitySharedManager.java:999) ~[classes/:?]
  at org.apache.ignite.internal.processors.cache.CacheAffinitySharedManager.onCacheChangeRequest(CacheAffinitySharedManager.java:885) ~[classes/:?]
  at org.apache.ignite.internal.processors.cache.distributed.dht.preloader.GridDhtPartitionsExchangeFuture.onCacheChangeRequest(GridDhtPartitionsExchangeFuture.java:1449) [classes/:?]
  at org.apache.ignite.internal.processors.cache.distributed.dht.preloader.GridDhtPartitionsExchangeFuture.init(GridDhtPartitionsExchangeFuture.java:947) [classes/:?]
  at org.apache.ignite.internal.processors.cache.GridCachePartitionExchangeManager$ExchangeWorker.body0(GridCachePartitionExchangeManager.java:3379) [classes/:?]
  at org.apache.ignite.internal.processors.cache.GridCachePartitionExchangeManager$ExchangeWorker.body(GridCachePartitionExchangeManager.java:3201) [classes/:?]
  at org.apache.ignite.internal.util.worker.GridWorker.run(GridWorker.java:120) [classes/:?]
  at java.lang.Thread.run(Thread.java:748) [?:1.8.0_261]
```

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java [hunk 1]
```diff
@@ -2385,12 +2388,30 @@ private void onCacheStarted(GridCacheContext cacheCtx) throws IgniteCheckedExcep
...
+    /**
+     * Get formatted string with expire policy info.
+     *
+     * @param cacheCtx - grid cache context.
+     * @return formatted expire policy info.
+     */
+    private String buildExpirePolicyInfo(GridCacheContext cacheCtx) {
+        ExpiryPolicy expPlc = cacheCtx.expiry();
+
+        Duration dur = expPlc.getExpiryForCreation() != null ? expPlc.getExpiryForCreation() :
```
Comment (ivandasch) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java:2407]: Expiry policy can be null, FYI

**Review by ivandasch — CHANGES_REQUESTED** [review]
Tests with expiry policy fails due to NPE, see comments

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java (hunk 1 above)
Comment (ivandasch) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java:2407]: Sorry, I suppose issue is not here.
You should not log expiry policy like here.
Expiry policy contains 3 fields and all of them can be used.

So you must log all durations of expiry policy. I suggests log -1 if duration is null

Comment (ivandasch) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java:2407]: Thin clients use `org.apache.ignite.internal.processors.platform.cache.expiry.PlatformExpiryPolicy`
It contains field `org.apache.ignite.internal.processors.platform.cache.expiry.PlatformExpiryPolicy#DUR_UNCHANGED`
If thin client pass it, thin client handler convert it to `null`. In tests one of parameter is `ExpiryPolicy()` with all fields `DUR_UNCHANGED`. And this case cause NPE in your code

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java [hunk 2]
```diff
@@ -1546,6 +1551,36 @@ private long checkPoolStarvation(
...
+
+                    Duration dur;
+                    if (expPlc.getExpiryForCreation() != null)
+                        dur = expPlc.getExpiryForCreation();
+                    else if (expPlc.getExpiryForUpdate() != null)
+                        dur = expPlc.getExpiryForUpdate();
+                    else
+                        dur = expPlc.getExpiryForAccess();
+
+                    if (dur == null || dur.getTimeUnit() == null) return null;
```
Comment (zstan) [inline modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java:1576]: return need to be on a new line

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java [hunk 3]
```diff
@@ -1546,6 +1551,36 @@ private long checkPoolStarvation(
...
+
+                    return "{cache=" + cacheCtx.name() + ", duration=" + dur.getTimeUnit().toMillis(dur.getDurationAmount()) +
+                            "ms, isEagerTtl=" + cacheCtx.ttl().eagerTtlEnabled() + "}";
+                }).filter(s -> !Objects.isNull(s))
+                .collect(Collectors.joining(", "));
+
+        log.info(String.format("Expiration policy info for caches [%s]", r));
```
Comment (zstan) [inline modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java:1583]: do we need to print this message if previously return null and thus final String r = "" ?

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java [hunk 4]
```diff
@@ -1546,6 +1551,36 @@ private long checkPoolStarvation(
...
+                    else if (expPlc.getExpiryForUpdate() != null)
+                        dur = expPlc.getExpiryForUpdate();
+                    else
+                        dur = expPlc.getExpiryForAccess();
+
+                    if (dur == null || dur.getTimeUnit() == null) return null;
+
+                    return "{cache=" + cacheCtx.name() + ", duration=" + dur.getTimeUnit().toMillis(dur.getDurationAmount()) +
+                            "ms, isEagerTtl=" + cacheCtx.ttl().eagerTtlEnabled() + "}";
+                }).filter(s -> !Objects.isNull(s))
```
Comment (zstan) [inline modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java:1580]: Objects.notNull more accurate here

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java [hunk 5]
```diff
@@ -1546,6 +1551,36 @@ private long checkPoolStarvation(
...
+                    if (expPlc.getExpiryForCreation() != null)
+                        dur = expPlc.getExpiryForCreation();
+                    else if (expPlc.getExpiryForUpdate() != null)
+                        dur = expPlc.getExpiryForUpdate();
+                    else
+                        dur = expPlc.getExpiryForAccess();
+
+                    if (dur == null || dur.getTimeUnit() == null) return null;
+
+                    return "{cache=" + cacheCtx.name() + ", duration=" + dur.getTimeUnit().toMillis(dur.getDurationAmount()) +
```
Comment (zstan) [inline modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java:1578]: why do you use "{" braces ? check other logs format style.

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java [hunk 6]
```diff
@@ -1546,6 +1551,36 @@ private long checkPoolStarvation(
...
+    }
+
+    /**
+     * Acks expiration policy info by caches.
+     */
+    private void ackExpPlcInfoByCaches() {
+        final String r = ctx.cache().context().cacheContexts().stream()
+                .map(cacheCtx -> {
+                    ExpiryPolicy expPlc = cacheCtx.expiry();
+                    if (expPlc == null || expPlc instanceof EternalExpiryPolicy) return null;
```
Comment (zstan) [inline modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java:1566]: return need to be on a new line

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java [hunk 7]
```diff
@@ -1546,6 +1551,36 @@ private long checkPoolStarvation(
...
         startTimer.finishGlobalStage("Await exchange");
+
+        if (log.isInfoEnabled())
+            ackExpPlcInfoByCaches();
+    }
+
+    /**
+     * Acks expiration policy info by caches.
+     */
+    private void ackExpPlcInfoByCaches() {
```
Comment (zstan) [inline modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java:1562]: func name confusing, change it for more informative.
Also we need tests checks correct logging all cache expiration variations.

**Pull request closed without merging**